openmc/model/surface_composite.py

```diff
@@ -621,3 +625,323 @@ def __init__(self, x0=0., y0=0., z0=0., r2=1., up=True, **kwargs):
+
+    def __init__(self, points, basis='rz'):
+        check_value('basis', basis, ('xy', 'yz', 'xz', 'rz'))
+        self._basis = basis
+        points = np.asarray(points, dtype=float)
+        check_iterable_type('points', points, float, min_depth=2, max_depth=2)
+        check_length('points', points[0, :], 2, 2)
+
+        # If the last point is the same as the first, remove it and make sure
+        # there are still at least 3 points for a valid polygon.
+        if np.allclose(points[0, :], points[-1, :]):
+            points = points[:-1, :]
+        check_length('points', points, 3)
+
+        # Order the points counter-clockwise (necessary for offset method)
+        self._points = self._make_ccw(points)
+
+        # Create a triangulation of the points.
+        self._tri = Delaunay(self._points, qhull_options='QJ')
+
+        # Decompose the polygon into groups of simplices forming convex subsets
+        # and get the sets of (surface, operator) pairs defining the polygon
+        self._surfsets = self._decompose_polygon_into_convex_sets()
+
+        # Set surface names as required by CompositeSurface protocol
+        surfnames = []
+        i = 0
+        for surfset in self._surfsets:
+            for surf, op, on_boundary in surfset:
+                if on_boundary:
+                    setattr(self, f'surface_{i}', surf)
+                    surfnames.append(f'surface_{i}')
+                    i += 1
+        self._surfnames = tuple(surfnames)
+
+        # Generate a list of regions whose union represents the polygon.
+        regions = []
+        for surfs_ops in self._surfsets:
+            regions.append([getattr(surf, op)() for surf, op, _ in surfs_ops])
+        self._regions = [openmc.Intersection(regs) for regs in regions]
+
+        # Create the union of all the convex subsets
+        self._region = openmc.Union(self._regions)
+
+    def __neg__(self):
+        return self._region
+
+    def __pos__(self):
+        return ~self._region
+
+    @property
+    def _surface_names(self):
+        return self._surfnames
+
+    @CompositeSurface.boundary_type.setter
+    def boundary_type(self, boundary_type):
+        if boundary_type != 'transmission':
+            warnings.warn("Setting boundary_type to a value other than "
+                          "'transmission' on Polygon composite surfaces can "
+                          "result in unintended behavior. Please use the "
+                          "regions property of the Polygon to generate "
+                          "individual openmc.Cell objects to avoid unwanted "
+                          "behavior.")
+        for name in self._surface_names:
+            getattr(self, name).boundary_type = boundary_type
+
+    @property
+    def points(self):
+        return self._points
+
+    @property
+    def basis(self):
+        return self._basis
+
+    @property
+    def _normals(self):
+        """Generate the outward normal unit vectors for the polygon."""
```

Comment:
No, but I can add a brief explanation if it helps. Essentially if we assume the points are ordered counter-clockwise then this simply rotates the vectors that point from one point to the next by -90 degrees about the z axis (if the basis is 'xy'), but the algorithm is the same regardless of the basis so that doesn't matter.

Reply:
Yes, for posterity, can you add a comment in here explaining what is done?